File: Number-of-Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int: 
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
        
                if(grid[i][j]=="1"):
                    
                    count+=1
                    self.helperDFS(grid,i,j)
        return count

    def helperDFS(self, grid,i,j):
        
        if(i<0 or j <0 or i >= len(grid) or j >= len(grid[0])): return 
        
        if(grid[i][j]=="0"): return
        if(grid[i][j]=="1"):
            grid[i][j]="0"
            
            self.helperDFS(grid,i+1,j)
            self.helperDFS(grid,i-1,j)
            self.helperDFS(grid,i,j+1)
            self.helperDFS(grid,i,j-1)
        
        
        

File: Number-of-Islands.java
class Solution {
    public int numIslands(char[][] grid) {
        if(grid==null || grid.length==0 || grid[0].length==0){return 0;}
        int count = 0;
        for(int i = 0; i < grid.length;i++)
        {
            for(int j = 0; j < grid[0].length; j++)
            {

                if(grid[i][j]=='1'){
                    
                    count+=1;
                    dfs(grid,i,j);
                }
                
            }
        }    
        return count;
    }
    
    void dfs(char[][] grid, int x,int y)
    {
        int length = grid.length;
        int height = grid[0].length;
        
        if(x<0 || y<0 || x>=length || y>=height || grid[x][y]=='0')
        {
            return;
        }
        grid[x][y] = '0';
        dfs(grid,x-1,y);
        dfs(grid,x,y-1);
        dfs(grid,x+1,y);
        dfs(grid,x,y+1);
        
    }
}
